compiler/rustc_const_eval/src/interpret/place.rs

```diff
@@ -419,8 +419,12 @@ where
         &self,
         val: &ImmTy<'tcx, M::Provenance>,
     ) -> InterpResult<'tcx, MPlaceTy<'tcx, M::Provenance>> {
+        let pointer_type = match val.layout.ty.kind() {
+            ty::Pat(base, _) => *base,
+            _ => val.layout.ty,
+        };
         let pointee_type =
-            val.layout.ty.builtin_deref(true).expect("`ref_to_mplace` called on non-ptr type");
+            pointer_type.builtin_deref(true).expect("`ref_to_mplace` called on non-ptr type");
```

Comment:
cc @rust-lang/miri some interpreter changes for making it easier for miri to handle pattern types.

Reply:
So can you explain what's happening here and why?

We had a refactor a while ago to ensure that in MIR, only actually primitive pointers appear -- no `Box`, in particular. This seems to regress that to re-introduce the mess of pointers masquerading under other types such as pattern types.